### PR TITLE
Bumping flash attention version to 2.6.3 and adding option for softcap in attention and lm_head logits.

### PR DESCRIPTION
**Pull request by ShashankMosaicML**
This PR bumps flash attention version to 2.6.3.
It also adds the option for softcap in attention and lm_head logits, to allow Gemma-like models. The config names are same as the huggingface names here: https://github.com/huggingface/transformers/blob/96a074fa7e2c04b904f72d9e827398d4c5f90f25/src/transformers/models/gemma2/modeling_gemma2.py#L371

Loss curves for attn softcapping vs no attn softcapping:
Screenshot 2024-09-23 at 10 36 14 AM

MFU curves for attn softcapping vs no attn softcapping:
Screenshot 2024-09-23 at 10 35 44 AM

### CODE REVIEW

File: tests/models/layers/test_flash_attn.py [hunk 1]
```diff
@@ -334,5 +335,99 @@ def gen_bias():
...
+    _assert_approx_equal(output_1, output_2)
+    assert (query_2.grad is not None) and (query_1.grad is not None)
+    _assert_approx_equal(query_1.grad, query_2.grad)
+    assert (key_2.grad is not None) and (key_1.grad is not None)
+    _assert_approx_equal(key_1.grad, key_2.grad)
+    assert (value_2.grad is not None) and (value_1.grad is not None)
+    _assert_approx_equal(value_1.grad, value_2.grad)
+
+
+def _assert_approx_equal(
```
Comment (dakinggg) [inline tests/models/layers/test_flash_attn.py:425]: is there a reason you implemented this yourself instead of using torch's built in functions?

**Review by dakinggg — APPROVED** [review]

**Review by dakinggg — APPROVED** [review]
